File: src/main/java/com/teamwizardry/librarianlib/api/gui/components/ComponentCenterAlign.java
package com.teamwizardry.librarianlib.api.gui.components;

import com.teamwizardry.librarianlib.api.gui.GuiComponent;
import com.teamwizardry.librarianlib.math.BoundingBox2D;
import com.teamwizardry.librarianlib.math.Vec2;

public class ComponentCenterAlign extends GuiComponent<ComponentCenterAlign> {

	public boolean centerHorizontal, centerVertical;
	
	public ComponentCenterAlign(int posX, int posY, boolean horizontal, boolean vertical) {
		super(posX, posY);
		centerHorizontal = horizontal;
		centerVertical = vertical;
	}

	@Override
	public void drawComponent(Vec2 mousePos, float partialTicks) {
		// noop
	}
	
	@Override
	public void draw(Vec2 mousePos, float partialTicks) {
		if(centerHorizontal || centerVertical) {
			for (GuiComponent<?> component : components) {
				Vec2 compPos = component.getPos();
				BoundingBox2D bb = component.getLogicalSize();
				Vec2 posOffsetFromBB = compPos.sub(bb.min);
				Vec2 centerPos = bb.max.sub(bb.min).mul(1f/2f).sub(posOffsetFromBB);
				Vec2 adjustedPos = centerPos.mul(-1);
				if(!centerHorizontal)
					adjustedPos = adjustedPos.setX(compPos.x);
				if(!centerVertical)
					adjustedPos = adjustedPos.setY(compPos.y);
				component.setPos(adjustedPos);
			}
		}
		
		super.draw(mousePos, partialTicks);
	}

}

File: src/main/java/com/teamwizardry/librarianlib/features/gui/components/ComponentCenterAlign.kt
package com.teamwizardry.librarianlib.features.gui.components

import com.teamwizardry.librarianlib.features.gui.component.GuiComponent
import com.teamwizardry.librarianlib.features.kotlin.div
import com.teamwizardry.librarianlib.features.kotlin.minus
import com.teamwizardry.librarianlib.features.kotlin.plus
import com.teamwizardry.librarianlib.features.math.Vec2d

/**
 * ## Facade equivalent: [StackLayout][com.teamwizardry.librarianlib.features.facade.layout.StackLayout]
 */
@Deprecated("As of version 4.20 this has been superseded by Facade")
class ComponentCenterAlign(posX: Int, posY: Int, var centerHorizontal: Boolean, var centerVertical: Boolean) : GuiComponent(posX, posY) {

    override fun drawComponent(mousePos: Vec2d, partialTicks: Float) {
        if (centerHorizontal || centerVertical) {
            for (component in children) {
                val componentCenter = component.transformToParentContext(component.size / 2)
                var adjustedPos = (size / 2) - componentCenter
                if (!centerHorizontal)
                    adjustedPos = adjustedPos.setX(0.0)
                if (!centerVertical)
                    adjustedPos = adjustedPos.setY(0.0)
                component.pos += adjustedPos
            }
        }
    }

}
